### PR TITLE
ReflectionTypeLoadException is thrown when assembly has dependencies

### PR DESCRIPTION
**Pull request by ar7z1**
In this PR I've fixed the problem @Jens-H-Eriksen described in #98.
As I wrote in [this comment](https://github.com/OctopusDeploy/OctoPack/pull/98#issuecomment-337477156) I've got the same issue.

@zentron clearly [described](https://github.com/OctopusDeploy/OctoPack/pull/98#issuecomment-331016705) the issue - `Assembly.Load(Byte[])` doesn't load assembly dependencies:
>By loading the Assembly via Assembly.Load(File.ReadAllBytes(path)) the process has no way of knowing where any dependent assemblies are located. It can try the GAC or the local path where the process is being executed, but in this sample case where it is trying to load types from Calamari.exe it is unable to find or load the Octostache or Nuget.Core dlls contained in the directory it was read from.

[Best Practices for Assembly Loading](https://docs.microsoft.com/en-us/dotnet/framework/deployment/best-practices-for-assembly-loading) has a lot more information about assembly loading procedure.

First of all, in this PR I've modified tests to reproduce the issue:
1. Added two dependent projects: `OctoPack.Tests.SampleDllDependency` and `OctoPack.Tests.SampleExeDependency`. As you can guess, one has `.dll` extension, another - `.exe`.
2. Added them as dependencies to the old sample projects (`OctoPack.Tests.NonGitVersionAssembly` and `OctoPack.Tests.SampleGitVersionAssembly`).
3. Dropped from `OctoPack.Tests` references to `OctoPack.Tests.NonGitVersionAssembly` and `OctoPack.Tests.SampleGitVersionAssembly`. Old tests didn't test real usage because `Assembly.Load` called for referenced assemblies.
4. Moved `GetNuGetVersionFromGitVersionInformation` method from [GetAssemblyVersionInfo.cs](https://github.com/OctopusDeploy/OctoPack/pull/99/files#diff-18d624ecbafcdc5effbf6b3f376cd398L133) to [AssemblyExtensions.cs](https://github.com/OctopusDeploy/OctoPack/pull/99/files#diff-7b14147167ee9cfb17070e0cff2f7842R26), made it public and called it from tests.
5. And finally `AssemblyExtensionsTests` have failed.

After that, I've added a handler for [AssemblyResolve](https://msdn.microsoft.com/en-us/library/system.appdomain.assemblyresolve.aspx?f=255&MSPPError=-2147217396) event. That handler tries to load an unknown assembly from path passed to `GetNuGetVersionFromGitVersionInformation` method.

### CODE REVIEW

File: source/OctoPack.Tasks/GetAssemblyVersionInfo.cs [hunk 1]
```diff
@@ -128,18 +126,6 @@ private TaskItem UseAssemblyVersion(FileVersionInfo info)
                 {"Version", assemblyVersionInfo},
             });
         }
-
-
-        private static string GetNuGetVersionFromGitVersionInformation(string path)
```
Comment (ar7z1) [inline source/OctoPack.Tasks/GetAssemblyVersionInfo.cs]: Moved to [AssemblyExtentions.cs](https://github.com/OctopusDeploy/OctoPack/pull/99/files#diff-7b14147167ee9cfb17070e0cff2f7842).

File: source/OctoPack.Tests.NonGitVersionAssembly/ClassInAssemblyWhereGitVersionIsNotUsed.cs [hunk 2]
```diff
@@ -1,12 +0,0 @@
-﻿using System;
-using System.Collections.Generic;
-using System.Linq;
-using System.Text;
-using System.Threading.Tasks;
-
-namespace OctoPack.Tests.NonGitVersionAssembly
-{
-    public class ClassInAssemblyWhereGitVersionIsNotUsed
```
Comment (ar7z1) [inline source/OctoPack.Tests.NonGitVersionAssembly/ClassInAssemblyWhereGitVersionIsNotUsed.cs]: Renamed to [ClassInAssemblyWhereGitVersionIsNotUsed1](https://github.com/OctopusDeploy/OctoPack/pull/99/files#diff-a53d4110ad6b55bf952277fabf61e0bd)

File: source/OctoPack.Tests.NonGitVersionAssembly/OctoPack.Tests.NonGitVersionAssembly.csproj [hunk 3]
```diff
@@ -33,17 +33,22 @@
   <ItemGroup>
     <Reference Include="System" />
     <Reference Include="System.Core" />
-    <Reference Include="System.Xml.Linq" />
```
Comment (ar7z1) [inline source/OctoPack.Tests.NonGitVersionAssembly/OctoPack.Tests.NonGitVersionAssembly.csproj]: Unused references.

File: source/OctoPack.Tests/Tasks/AssemblyExtensionsTests.cs [hunk 4]
```diff
@@ -1,25 +1,34 @@
...
+            var assemblyPath = Path.Combine(currentAssemblyPath, "..", "..", "..", "..", name, "bin", configuration, $"{name}.dll");
+            return new FileInfo(assemblyPath).FullName;
         }
 
         [Test]
         public void AssertAssemblyVersion_WhereNoGitVersionProperty_ReturnsNull()
         {
-            string gitversion = Assembly.GetAssembly(typeof(ClassInAssemblyWhereGitVersionIsNotUsed)).GetNugetVersionFromGitVersionInformation();
+            var assemblyPath = GetAssemblyFullPath("OctoPack.Tests.NonGitVersionAssembly");
+            var gitversion = AssemblyExtensions.GetNuGetVersionFromGitVersionInformation(assemblyPath);
```
Comment (ar7z1) [inline source/OctoPack.Tests/Tasks/AssemblyExtensionsTests.cs]: Now tests use the same code to load assembly as MSBuild task.

**Comment by zentron** [comment]
Thanks for the PR with tests @ar7z1 . I was able to reproduce it consistently with these code changes and the fix seems to do the job.

**Comment by ar7z1** [comment]
@zentron Thank you for the review! Can you publish a new version of NuGet package?

**Comment by zentron** [comment]
Way ahead of ya @ar7z1  :)
image

**Comment by ar7z1** [comment]
@zentron Great work! Thank you! :-)

**Pull request merged**